sound/soc/sof/intel/hda-dai-ops.c

```diff
@@ -524,16 +524,25 @@ static const struct hda_dai_widget_dma_ops hda_ipc4_chain_dma_ops = {
 static int hda_ipc3_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
 				 struct snd_pcm_substream *substream, int cmd)
 {
+	struct hdac_ext_stream *hext_stream = hda_get_hext_stream(sdev, cpu_dai, substream);
 	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(cpu_dai, substream->stream);
 
 	switch (cmd) {
 	case SNDRV_PCM_TRIGGER_SUSPEND:
 	case SNDRV_PCM_TRIGGER_STOP:
 	{
 		struct snd_sof_dai_config_data data = { 0 };
+		int ret;
 
 		data.dai_data = DMA_CHAN_INVALID;
-		return hda_dai_config(w, SOF_DAI_CONFIG_FLAGS_HW_FREE, &data);
+		ret = hda_dai_config(w, SOF_DAI_CONFIG_FLAGS_HW_FREE, &data);
+		if (ret < 0)
+			return ret;
+
+		if (cmd == SNDRV_PCM_TRIGGER_STOP)
+			return hda_link_dma_cleanup(substream, hext_stream, cpu_dai);
```

Comment:
I don't understand how this is compatible with the previous direction to keep the stream tag in  https://github.com/thesofproject/linux/commit/1bf83fa6654ce8959948878aad14a6db586125b8 whose commit message included 

"In the case of repeated start/stop without involving hw_free, the stream
tag needs to be preserved for the subsequent starts. So, skip performing
the DMA clean up during stop and handle it only during suspend or
hw_free.
"

??

Reply:
@plbossart that commit was a mistake for IPC3. In the case of IPC3, the post_trigger op frees clears the DMA channel ID and send the DAI_CONFIG to clear the DMA iD in the firmware as well. So we must clean up the link DMA set up so that we can re-assign and do the DAI_CONFIG again during prepare in the case of repeasted hw_params without hw_free.

Reply:
ah ok, the commit https://github.com/thesofproject/linux/commit/1bf83fa6654ce8959948878aad14a6db586125b8 was for IPC4, makes sense now. Thanks for the clarification.